### ISSUE TITLE
containerized controller manager doesn't generate secrets w/ default configuration

### ISSUE THREAD

**Issue by dalanlan** [kubernetes/kubernetes#16965]
As specified in #13557, containerized kubelet has a problem when it comes to secrets thing. #13791 (merged on Sep. 18th) should make it right.
Yet I built one from HEAD on Oct 16th, and here is what i got.
```
$ kubectl get secrets --all-namespaces
NAMESPACE   NAME      TYPE      DATA
$ kubectl get serviceaccount --all-namespaces
NAMESPACE   NAME      SECRETS
default     default   0
$ kubectl exec busybox -- ls /var/run/secrets
ls: /var/run/secrets: No such file or directory
error: Error executing remote command: Error executing command in container: Error executing in Docker Container: 1
```
Am i missing something?

/cc @fgrzadkowski  @pmorie

**Comment by dalanlan** [comment]
What's expected in my mind
```
$ kubectl exec busybox -- ls /var/run/secrets
kubernetes.io
$ kubectl get secrets
NAME                  TYPE                                  DATA
default-token-towes   kubernetes.io/service-account-token   2
$ kubectl get serviceaccount
NAME      SECRETS
default   1

```

**Comment by luxas** [comment]
Yes, it should be like that, and it works when using `hack/local-up-cluster.sh` from HEAD.
When it works with local-up-cluster: 
```
# running on host
mount
tmpfs on /var/lib/kubelet/pods/6160ea3c-8538-11e5-855f-f0761c62f136/volumes/kubernetes.io~secret/default-token-1cmtw type tmpfs (rw,relatime)

kubectl exec busybox -- mount
tmpfs on /var/run/secrets/kubernetes.io/serviceaccount type tmpfs (ro,relatime)

ls /var/lib/kubelet/pods/<busybox_pod_id>/volumes
kubernetes.io~secret
```


But, not when running `kubelet` under docker.
When executing the commands above using dockerized kubelet `v1.2.0-alpha.3` with `--privileged` and `--containerized`, it doesn't return anything.
I'll try to set `-v=3` and see if kubelet says something new. Now, no errors regarding mounts is thrown.

**Comment by luxas** [comment]
@dalanlan I've found an interesting thing. It is working to follow the guide: https://github.com/kubernetes/kubernetes/tree/master/docs/user-guide/secrets

I got secrets working with dockerized kubelet after #13791. Prior to the patch, secrets didn't mount.
Output:
```
# running on host
mount
tmpfs on /var/lib/kubelet/pods/<pod_id>/volumes/kubernetes.io~secret/secret-volume type tmpfs (rw,relatime)
tmpfs on /var/lib/kubelet/pods/<pod_id>/volumes/kubernetes.io~secret/secret-volume type tmpfs (rw,relatime)

kubectl exec busybox -- mount
tmpfs on /etc/secret-volume type tmpfs (ro,relatime)

ls /var/lib/kubelet/pods/<pod_id>/volumes
kubernetes.io~secret
```
(The reason why `tmpfs` is mounted twice, is probably because I tried this some times)

So, according to this, it seems like dockerized kubelet just don't bother to create the `default serviceAccount secret`. Because @fgrzadkowski made it possible, kubelet should be capable. 
I'll dig into the code and see if I find something.

**Comment by luxas** [comment]
@dalanlan @fgrzadkowski @resouer @pmorie I got this working! :smile:

I've been searching for this problem today and now I've found the solution.
First off, this hadn't been possible without: #13791

The problem was the `controller-manager`, because it includes the **serviceaccounts controller**
It didn't know about the self-signed certs, so it didn't inject the serviceaccount secret volumes to every container. And it didn't either create the default secret for serviceAccounts (not even if one manually created a serviceAccount)

What can we do about this? Earlier today I found out that if no args is passed to apiserver, it self-signs two files: `/var/run/kubernetes/apiserver.key` and `/var/run/kubernetes/apiserver.crt`.
The serviceaccounts controller wants these certs.

So one solution and the simpliest one:
master.json
```
...
"containers":[
    {
      "name": "controller-manager",
      "image": "gcr.io/google_containers/hyperkube:v1.2.0-alpha.3",
      "command": [
              "/hyperkube",
              "controller-manager",
              "--master=127.0.0.1:8080",
              "--service-account-private-key-file=/var/run/kubernetes/apiserver.key",
              "--root-ca-file=/var/run/kubernetes/apiserver.crt",
              "--v=2"
        ],
        "volumeMounts":[{
          "name": "kubecerts",
          "mountPath": "/var/run/kubernetes"
        }]
    },
    {
      "name": "apiserver",
      "image": "gcr.io/google_containers/hyperkube:v1.2.0-alpha.3",
      "command": [
              "/hyperkube",
              "apiserver",
              "--admission-control=NamespaceLifecycle,NamespaceAutoProvision,LimitRanger,ServiceAccount,ResourceQuota",
              "--service-cluster-ip-range=10.0.0.1/24",
              "--address=127.0.0.1",
              "--etcd-servers=http://127.0.0.1:4001",
              "--cluster-name=kubernetes",
              "--v=2"
        ],
        "volumeMounts":[{
          "name": "kubecerts",
          "mountPath": "/var/run/kubernetes"
        }]
    },
    {
      "name": "scheduler",
      "image": "gcr.io/google_containers/hyperkube:v1.2.0-alpha.3",
      "command": [
              "/hyperkube",
              "scheduler",
              "--master=127.0.0.1:8080",
              "--v=2"
        ]
    }
  ],
"volumes": [
    {
      "name": "kubecerts",
      "emptyDir": {}
    }
  ]
...
```

Alternatives:
 - Use `emptyDir`. `apiserver` self-signs the certs and `controller-manager` uses them. In-memory.
 - Use an empty `hostPath`. `apiserver` self-signs the certs and `controller-manager` uses them. Stored on the master's filesystem.
 - Use a prepopulated dir with `hostPath` (or another volume plugin). Specify `apiserver` and 'controller-manager` flags which points to already-made certs.

I don't exactly know which `--admission-control=` arguments that should be there, but at least `ServiceAccount` is needed for `/var/run/secrets/kubernetes.io` to be present.

Results:
```
kubectl get serviceaccounts
NAME      SECRETS   AGE
default   1         2m

kubectl get secrets
NAME                  TYPE                                  DATA      AGE
default-token-1b0wy   kubernetes.io/service-account-token   2         2m

kubectl get po/busybox -o yaml
...
spec:
  containers:
  - command:
    - sleep
    - "3600"
    image: busybox
    imagePullPolicy: IfNotPresent
    name: busybox
    resources: {}
    terminationMessagePath: /dev/termination-log
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: default-token-kq30c
      readOnly: true
  dnsPolicy: ClusterFirst
  nodeName: 127.0.0.1
  restartPolicy: Always
  securityContext: {}
  serviceAccount: default
  serviceAccountName: default
  terminationGracePeriodSeconds: 30
  volumes:
  - name: default-token-kq30c
    secret:
      secretName: default-token-kq30c
...

kubectl exec busybox -- ls /var/run/secrets
kubernetes.io

kubectl exec -it busybox -- ls /var/run/secrets/kubernetes.io/serviceaccount
ca.crt  token
```

Another thing, it's quite messy to edit the `master(-multi).json` file in an already built `hyperkube` image. The quick and dirty debug solution is `-v $(pwd)/master.json:/etc/kubernetes/manifests/master.json`
But that is an other question.

WDYT?

**Comment by pmorie** [comment]
Glad to hear you got is working @luxas.  This is going to get a lot easier once the mount propagation control lands in docker.  Right now we have to do some hacks using nsenter to mount and write data into the host's mount namespace when the kubelet is containerized so that the volumes are visible to other containers.  That's the issue you were running into, but as you say, #13791 solves this by nsentering the host's mount ns and writing the data using echo.

**Comment by pmorie** [comment]
@dalanlan Check out `local-up-cluster.sh` -- you need to use `--containerized` and some specific bind-mounts for kubelet to function properly in a container.

**Comment by dalanlan** [comment]
@pmorie I did. The upstart params for containerized kubelet i used are listed following
```
docker run \
    --volume=/:/rootfs:ro \
    --volume=/sys:/sys:ro \
    --volume=/dev:/dev \
    --volume=/var/lib/docker/:/var/lib/docker:rw \
    --volume=/var/lib/kubelet/:/var/lib/kubelet:rw \
    --volume=/var/run:/var/run:rw \
    --net=host \
    --pid=host \
    --privileged=true \
    -d \
    gcr.io/google_containers/hyperkube:<from-head> \
    /hyperkube kubelet --containerized --hostname-override="127.0.0.1" --address="0.0.0.0" --api-servers=http://localhost:8080 --config=/etc/kubernetes/manifests
```

**Comment by dalanlan** [comment]
I had done several experiments w/ `--service-account-private-key-file` & `--root-ca-file` filled for controller manager months ago, and it should work indeed. Yet i expect secrets should be generated automatically (at least it does on bare-metal ubuntu now) w/o passing these flags (correct me if i'm wrong). Maybe that's another issue with controller manager.

**Comment by dalanlan** [comment]
@pmorie  Could you clarify this for me? Is it expected that w/ [these configuration](https://github.com/kubernetes/kubernetes/issues/16965#issuecomment-154769226), secrets would be generated automatically (just as directly run on bare-metal ubuntu) or not?

**Comment by dalanlan** [comment]
Oh, guess what... I could labour a delusion of this whole thing. There is no way SA controller would generate secrets automatically.
Closing it. And sorry for the chaos.

**Comment by luxas** [comment]
@dalanlan Should we open a new issue for enabling secret functionality by default in hyperkube?

IMO, `secrets` and `serviceAccounts` are quite a big part of k8s, and the `k8s-in-docker` setup will suffer a lot if these aren't default.

**Comment by fgrzadkowski** [comment]
@luxas secrets are working in k8s-in-docker setup. Service accounts are not. We are now working on a proposal how to simplify the whole procedure. It will also address things like certificates and service accounts.

**Comment by luxas** [comment]
@fgrzadkowski Oh sorry, I meant the combination of those, that `secrets` aren't created when a `ServiceAccount` is. I do know that secrets were fixed in your PR: #13791

However, I did solve it here: https://github.com/kubernetes/kubernetes/issues/16965#issuecomment-154740451 
My PR is here: #17213
And maybe the discussion continues in this PR after the merge: #17262

**Comment by dalanlan** [comment]
> @luxas secrets are working in k8s-in-docker setup.

@fgrzadkowski  If that's the case, it's an issue indeed. As i pointed out in the [comment](https://github.com/kubernetes/kubernetes/issues/16965#issue-115661647), i could not get any secrets w/ default configuration (by which i mean, no additional flag apart from HEAD now)

**Comment by fgrzadkowski** [comment]
@dalanlan Secrets and service accounts are not the same thing. Please create a secret manually and mount it to your pod to verify if that working or not. I believe it should be ok. Service accounts tokens are indeed not generated in the default setup and there is a PR in-flight that fixes it.

Please reopen if you can reproduce that a secrets are not properly mounted/written when you create them manually and attach them to your pod (please see http://kubernetes.io/v1.0/docs/user-guide/secrets.html for more details).

**Comment by dalanlan** [comment]
> Service accounts tokens are indeed not generated in the default setup and there is a PR in-flight that fixes it.

That's expected then. Thank you for the clarification.

**Comment by stevenmccord** [comment]
@fgrzadkowski It looks like you have solved this issue, but I am not sure how I pull the latest docker images?  When I try to pull anything with 1.2 it looks like it isn't tagged.  How do I get the latest hyperkube image?  Thanks!
